### PR TITLE
Add gauge metric for snapshot controller operations in flight

### LINKED ISSUE #434: Add metric for detecting metrics operation leaks in snapshot controller
It is possible that operation leaks can occur if logic changes happen in the snapshot controller. We should add a metric to count how many snapshot operations are cached. If this count does not decrease after any operation is finished, we have an operation cache leak.

_Originally mentioned by @msau42 in https://github.com/kubernetes-csi/external-snapshotter/pull/409#discussion_r532979019_

### PR DESCRIPTION
**Pull request by ggriffiths**
Signed-off-by: Grant Griffiths <grant@portworx.com>

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Adds an operations in flight metrics to find any leaky operations in our metrics implementation. If a snapshot is not in progress, and in flight metrics are >0, we have a leak. I've done some basic testing with this metric and haven't found any leaked operations yet.

**Which issue(s) this PR fixes**:
Fixes #434 

**Special notes for your reviewer**:
I decided to change the sync.Map to be a normal map with a RWMutex. This is because sync.Map does not allow the user to find its length (golang/go issue 20680).

The original issue calls to check for any built-in way to determine if we have leaked operations. Except for some pprof memory profiling, I didn't see an easy way to do this via go tooling. This metric will allow us to easily tell if there's a leak.

**Does this PR introduce a user-facing change?**:
```release-note
Adds an in-operations_in_flight metric for determining the number of snapshot operations in progress.
```

### CODE REVIEW

File: pkg/metrics/metrics.go [hunk 1]
```diff
@@ -261,6 +280,29 @@ func (opMgr *operationMetricsManager) init() {
 		[]string{labelDriverName, labelOperationName, labelSnapshotType, labelOperationStatus},
 	)
 	opMgr.registry.MustRegister(opMgr.opLatencyMetrics)
+	opMgr.opInFlight = k8smetrics.NewGauge(
+		&k8smetrics.GaugeOpts{
+			Subsystem: subSystem,
+			Name:      operationInFlightName,
+			Help:      operationInFlightHelpMsg,
+		},
```
Comment (yuxiangqian) [inline pkg/metrics/metrics.go:288]: it might be helpful to also add labels like DriverName

File: pkg/metrics/metrics_test.go [hunk 2]
```diff
@@ -478,6 +478,87 @@ snapshot_controller_operation_total_seconds_count{driver_name="driver5",operatio
...
+
+	if err := verifyInFlightMetric(`snapshot_controller_operations_in_flight 2`, srvAddr); err != nil {
+		t.Errorf("failed testing [%v]", err)
+	}
+
+	//  Record, should be down to 1
+	mgr.RecordMetrics(opKey, nil, "driver")
+	time.Sleep(250 * time.Millisecond)
+
+	if err := verifyInFlightMetric(`# HELP snapshot_controller_operations_in_flight [ALPHA] Total number of operations in flight
```
Comment (yuxiangqian) [inline pkg/metrics/metrics_test.go:518]: why this string is different from line 510 & 536?

File: pkg/metrics/metrics_test.go [hunk 3]
```diff
@@ -478,6 +478,87 @@ snapshot_controller_operation_total_seconds_count{driver_name="driver5",operatio
...
+	mgr.OperationStart(opKey, opVal)
+	time.Sleep(250 * time.Millisecond)
+
+	if err := verifyInFlightMetric(`snapshot_controller_operations_in_flight 2`, srvAddr); err != nil {
+		t.Errorf("failed testing [%v]", err)
+	}
+
+	//  Record, should be down to 1
+	mgr.RecordMetrics(opKey, nil, "driver")
+	time.Sleep(250 * time.Millisecond)
```
Comment (yuxiangqian) [inline pkg/metrics/metrics_test.go:516]: this can cause flaky tests easily, can we avoid it?

File: pkg/metrics/metrics_test.go [hunk 4]
```diff
@@ -613,6 +695,14 @@ func containsMetrics(expectedMfs, gotMfs []*cmg.MetricFamily) bool {
 						fmt.Println("difference in counter values")
 						continue
 					}
+
+				case got.Metric[i].Gauge != nil && expected.Metric[j].Gauge != nil:
```
Comment (yuxiangqian) [inline pkg/metrics/metrics_test.go:699]: I do not see "expected" change?

File: pkg/metrics/metrics.go [hunk 5]
```diff
@@ -152,49 +158,58 @@ type operationMetricsManager struct {
...
 	return mgr
 }
 
 // OperationStart starts a new operation
 func (opMgr *operationMetricsManager) OperationStart(key OperationKey, val OperationValue) {
+	opMgr.mu.Lock()
+	defer opMgr.mu.Unlock()
+
 	val.startTime = time.Now()
-	opMgr.cache.LoadOrStore(key, val)
```
[the remark this answers is not in the document]
Comment (ggriffiths) [reply pkg/metrics/metrics.go:176]: Yes, unfortunately it is not possible to get the length of a sync.Map. There are ways to do this with map.Range, but it seemed expensive for every operation.

I will change OperationStart to preserve the existing behavior of `LoadOrStore`.

File: pkg/metrics/metrics_test.go (hunk 3 above)
Comment (ggriffiths) [reply pkg/metrics/metrics_test.go:516]: If you're referring to the time.Sleep, I think most of the existing tests depend on this already and I haven't noticed any flakes. I've increased this to 500 milliseconds for now.

File: pkg/metrics/metrics_test.go (hunk 4 above)
Comment (ggriffiths) [reply pkg/metrics/metrics_test.go:699]: What do you mean by this? I just changed one of the test expectations from 51 to 52 and it failed as expected. Am I missing something here?

File: pkg/metrics/metrics.go (hunk 1 above)
Comment (ggriffiths) [reply pkg/metrics/metrics.go:288]: The goal of this metric, per #434, is mainly for us to see if any leaks are happening. The metric counts all operations in transit regardless of driver name/labels.

There are some cases where we won't have driver name. i.e. in the scheduled go-routine and in DropOperation.

File: pkg/metrics/metrics_test.go (hunk 2 above)
Comment (ggriffiths) [reply pkg/metrics/metrics_test.go:518]: Fixed this

File: pkg/metrics/metrics.go [hunk 6]
```diff
@@ -246,7 +264,8 @@ func (opMgr *operationMetricsManager) recordCancelMetric(obj interface{}, key Op
 		val.SnapshotType,
 		string(SnapshotStatusTypeCancel),
 	).Observe(duration)
-	opMgr.cache.Delete(key)
+	delete(opMgr.cache, key)
+	opMgr.opInFlight.Set(float64(len(opMgr.cache)))
```
[the remark this answers is not in the document]
Comment (ggriffiths) [reply pkg/metrics/metrics.go]: Removed this

**Review by yuxiangqian — COMMENTED** [review]
nits and LGTM, thanks @ggriffiths

File: pkg/metrics/metrics_test.go (hunk 4 above)
Comment (ggriffiths) [reply pkg/metrics/metrics_test.go]: I spent some time trying to consolidate verifyInFlightMetric into verifyMetrics, but I found some unexpected test framework behavior that made it difficult to integrate the two. `verifyInFlightMetric` works well for this specific test.

Comment (ggriffiths) [reply pkg/metrics/metrics_test.go]: The main issues were the following:
1. `verifyMetrics` silently passes when an expected metric is not formatted correctly
2. `sortMfs` returns an empty slice always, which causes test validation to pass incorrectly
3. ran into some nil pointers when getting `sortMfs` to work with two different metric types.

I can open an issue to update the existing `verifyMetrics`. Was hoping to get this PR in with `verifyInFlightMetrics` for now.

**Comment by ggriffiths** [comment]
@yuxiangqian ready for review

**Comment by ggriffiths** [comment]
@yuxiangqian @xing-yang any additional work needed for this PR?

**Comment by xing-yang** [comment]
@yuxiangqian Do you have more comments on this PR?

**Review by yuxiangqian — COMMENTED** [review]
nit. and LGTM

File: pkg/metrics/metrics.go [hunk 7]
```diff
@@ -213,31 +229,34 @@ func (opMgr *operationMetricsManager) RecordMetrics(opKey OperationKey, opStatus
...
 		}
 	}
 
-	opMgr.cache.Delete(opKey)
+	delete(opMgr.cache, opKey)
+	opMgr.opInFlight.Set(float64(len(opMgr.cache)))
 }
 
 // recordCancelMetric records a metric for a create operation that hasn't finished
 func (opMgr *operationMetricsManager) recordCancelMetric(obj interface{}, key OperationKey, duration float64) {
```
[the remark this answers is not in the document]
Comment (ggriffiths) [reply pkg/metrics/metrics.go]: Addressed this

File: pkg/metrics/metrics.go [hunk 8]
```diff
@@ -152,49 +158,59 @@ type operationMetricsManager struct {
 	// ongoing operations.
 	// key is an Operation
 	// value is the timestamp of the start time of the operation
-	cache sync.Map
+	cache map[OperationKey]OperationValue
+
+	// mutex for protecting cache from concurrent access
+	mu sync.RWMutex
```
[the remark this answers is not in the document]
Comment (ggriffiths) [reply pkg/metrics/metrics.go]: Moved this to a sync.Mutex.

**Review by yuxiangqian — COMMENTED** [review]
/lgtm

**Pull request merged**